### ISSUE TITLE
vsock: error reading from backing stream

### ISSUE THREAD

**Issue by stefano-garzarella** [firecracker-microvm/firecracker#1751]
While developing the vhost-user-vsock application for cloud-hypervisor, I had an issue that happens also with the vsock device emulation: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1001

As @sboeuf suggested, I tried with Firecracker and I had a very similar issue:

```
2020-04-06T11:52:33.632842533 [anonymous-instance:ERROR:src/devices/src/virtio/vsock/csm/connection.rs:242] vsock: error reading from backing stream: lp=5201, pp=1031, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
```
After the error, the throughput goes to 0 and the peers are able to communicate only restarting the connection, but the error happens every time I tried to run `iperf-vsock` (guest connecting to the host)

## How to reproduce

I used [iperf-vsock](https://github.com/stefano-garzarella/iperf-vsock) to stress the vsock connection.

Kernel: hello-vmlinux.bin (Linux (none) 4.14.55-84.37.amzn2.x86_64 #1 SMP Wed Jul 25 18:47:15 UTC 2018 x86_64 Linux)
Firecracker: firecracker-v0.21.1-x86_64

### VM config
```json
{
  "boot-source": {
    "kernel_image_path": "./hello-vmlinux.bin",
    "boot_args": "console=ttyS0 reboot=k panic=1 pci=off"
  },
  "drives": [
    {
      "drive_id": "rootfs",
      "path_on_host": "rootfs.ext4",
      "is_root_device": true,
      "is_read_only": false
    }
  ],
  "machine-config": {
    "vcpu_count": 2,
    "mem_size_mib": 1024,
    "ht_enabled": false
  },
  "vsock": {
      "vsock_id": "1",
      "guest_cid": 4,
      "uds_path": "/tmp/vm4.vsock"
  }
}
```

### guest
```shell
~# mkdir /tmp     # iperf3 uses mkstemp(3)
~# iperf3 --vsock -c 2
Connecting to host 2, port 5201
[  5] local 4 port 1033 connected to 2 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  91.0 MBytes   763 Mbits/sec                  
2020-04-06T11:55:51.001682775 [anonymous-instance:ERROR:src/devices/src/virtio/vsock/csm/connection.rs:242] vsock: error reading from backing stream: lp=5201, pp=1033, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[  5]   1.00-2.00   sec   422 MBytes  3.54 Gbits/sec                  
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   513 MBytes   431 Mbits/sec                  sender
[  5]   0.00-10.01  sec   513 MBytes   430 Mbits/sec                  receiver

iperf Done.
```

## host
```shell
> ./iperf3 --vsock -s -B /tmp/vm4.vsock
-----------------------------------------------------------
Server listening on 5201
-----------------------------------------------------------
Accepted connection from 40:8c44:a220:e9bb:30b4:aa01::, port 43521
[  5] local 6d34:2e76:736f:636b:5f35:3230:3100:0 port 12148 connected to :: port 0
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   431 MBytes  3.61 Gbits/sec                  
[  5]   1.00-2.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   431 MBytes   361 Mbits/sec                  receiver
```

**Comment by stefano-garzarella** [comment]
Starting the `iperf-vsock` server in the guest and connecting from the host works, but if I use `-R` parameter, that means reverse mode (server sends, client receives) I hade the same behavior but a slightly different error message:

## guest

```shell
~# iperf3 --vsock -s
-----------------------------------------------------------
Server listening on 5201
-----------------------------------------------------------
Accepted connection from 2, port 1073741840
[  5] local 4 port 5201 connected to 2 port 1073741841
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   522 MBytes  4.38 Gbits/sec                  
2020-04-06T12:13:35.862177572 [anonymous-instance:WARN:src/devices/src/virtio/vsock/csm/connection.rs:288] vsock: error writing to local stream (lp=1073741841, pp=5201): TxBufFull
[  5]   1.00-2.00   sec   137 MBytes  1.15 Gbits/sec                  
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   659 MBytes   553 Mbits/sec                  sender
```

## host

```shell
> ./iperf3 --vsock -c /tmp/vm4.vsock -R
Connecting to host /tmp/vm4.vsock, port 5201
Reverse mode, remote host /tmp/vm4.vsock is sending
[  5] local :: port 0 connected to 6d34:2e76:736f:636b:: port 12148
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   522 MBytes  4.38 Gbits/sec                  
[  5]   1.00-2.00   sec   137 MBytes  1.15 Gbits/sec                  
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   659 MBytes   553 Mbits/sec                  sender
[  5]   0.00-10.00  sec   659 MBytes   553 Mbits/sec                  receiver

iperf Done.
```

**Comment by sandreim** [comment]
Hi @stefano-garzarella! Thanks for logging this. We'll get back to you after we look into it.

**Comment by dhrgit** [comment]
@stefano-garzarella I'm having trouble replicating the problem. Configurations that I've tried so far:
- Firecracker master HEAD and Firecracker v0.21.1
- iperf-vsock master HEAD
- guest kernel `hello-vmlinux.bin`
- host kernel 4.14.72 and 5.3.0 
- guest server + host client
- guest client + host server
- guest server + host client -R

All combinations seem to work fine, even though with a pretty low throughput. Haven't fiddled with buffer sizes yet though, which seems likely to be the cause.

Back to the errors, both of them seem pretty peculiar. In the first case, you get an unexpected EAGAIN / EWOULDBLOCK. That suggests the vsock back end received an EPOLLIN event for the unix socket, attempted to read data from it, yet no data was available. Seems weird, and I don't know what could cause this (i.e. EPOLLIN without any data available).

The second error seems even more bizarre (`TxBufFull`). This would suggest that one of the peers is not respecting flow control. I.e. either Firecracker or the guest vsock driver is mishandling vsock flow control.

Both situations require a more in-depth look, so I'd appreciate some more help with reproducing the issue. Thanks!

**Comment by stefano-garzarella** [comment]
@dhrgit thanks for taking a look!

> All combinations seem to work fine, even though with a pretty low throughput. Haven't fiddled with buffer sizes yet though, which seems likely to be the cause.

It is strange, for me it happens every time. The host shouldn't be important, anyway I'm on Fedora 31 (Linux 5.5.13-200.fc31.x86_64).

> Back to the errors, both of them seem pretty peculiar. In the first case, you get an unexpected EAGAIN / EWOULDBLOCK. That suggests the vsock back end received an EPOLLIN event for the unix socket, attempted to read data from it, yet still received EAGAIN. Seems weird, and I don't know what could cause this (i.e. EPOLLIN followed by EAGAIN).

At first, it seemed strange to me too, then I thought about these events:

- host's application sends X bytes ( X < buffer size) on the UDS
- FC receives EPOLLIN event and starts to call the backend callback to read the data
- before the FC's read(), host's application sends other Y bytes, new EPOLLIN event queued
- FC read()s X+Y bytes
- FC receives the EPOLLIN queued (can happen? or is the event canceled since the kernel buffer is empty? I should check better the epoll() implementation)

> The second error seems even more bizarre (`TxBufFull`). This would suggest that one of the peers is not respecting flow control. I.e. either Firecracker or the guest vsock driver is mishandling vsock flow control.

IIUC, in this case, the UDS buffer in the kernel is full (maybe because the host's application is slow).

Is FC telling the guest the available space in the UDS buffer or the space of the FC internal buffer?
(Sorry, I'm new in the FC code)

> Both situations require a more in-depth look, so I'd appreciate some more help with reproducing the issue. Thanks!

Sure, I'll try to help :-)
Let me know if you want me to run any specific test.

**Comment by sboeuf** [comment]
The important part to make sure you're both testing the with the same setup is the guest kernel version as there might have been some vsock patches that cause this behavior change.

**Comment by stefano-garzarella** [comment]
@sboeuf Yes, I agree.

I used `hello-vmlinux.bin (Linux 4.14.55-84.37.amzn2.x86_64)` (I think @dhrgit tried also this kernel).
Now I tried Linux 5.6.0 and it happens immediately (maybe because starting from Linux 5.4, the guest driver is faster).

Limiting the bitrate to 1 Gbps (`iperf3 --vsock -c 2 -b 1G` or `iperf3 --vsock -c /tmp/vm4.vsock -R -b 1G`) doesn't show the problem. Limiting the bitrate to 2 Gbps makes it appear after a while.

**Comment by dhrgit** [comment]
Ah, this is indeed bizarre: after I left iperf running for about 10 mins, I managed to reproduce the stalling error on 4.14. After that, it started to happen within seconds of starting iperf.

Similar for 5.6 - after it ran fine for a few minutes, it started failing consistently with a broken pipe error.

Alright, now I have something to investigate. I'll look into it and report back.

> Is FC telling the guest the available space in the UDS buffer or the space of the FC internal buffer?

It's the internal buffer state that drives flow control (`VsockConnection::tx_buf`). So, if that buffer gets overrun, it probably means that somebody's mishandling flow control.

**Comment by stefano-garzarella** [comment]
> Ah, this is indeed bizarre: after I left iperf running for about 10 mins, I managed to reproduce the stalling error on 4.14. After that, it started to happen within seconds of starting iperf.

Yes, it is bizarre...

> Similar for 5.6 - after it ran fine for a few minutes, it started failing consistently with a broken pipe error.
> 
> Alright, now I have something to investigate. I'll look into it and report back.

Thanks!

> 
> > Is FC telling the guest the available space in the UDS buffer or the space of the FC internal buffer?
> 
> It's the internal buffer state that drives flow control (`VsockConnection::tx_buf`). So, if that buffer gets overrun, it probably means that somebody's mishandling flow control.

Mmm, could be related to a wrap-around of peer_fwd_cnt and fwd_cnt?

**Comment by dhrgit** [comment]
Made some progress investigating the first issue (unexpected `EAGAIN`): it seems that we're getting some spurious `EPOLLIN` events, bundled with the expected `EPOLLOUT`s. Once the transfer starts, the vsock backend is driven to high TX load, and sometimes cannot write the entire available TX data to the Unix socket. That's when it allocates a `TxBuf`, and registers for `EPOLLOUT` events on the Unix socket. This is normal behavior.

However, it seems that, as far as I can tell randomly, `epoll_wait()` returns  `EPOLLOUT | EPOLLIN` instead of just `EPOLLOUT`. This would be ok if the host-side also wrote some data on the socket, buf as far as I can tell it doesn't. Firecracker then attempts to both flush the TX buffer, and read from the socket, to satisfy both `EPOLLOUT` and `EPOLLIN`. Only that reading fails with `EAGAIN`, which, together with the apparent randomness of occurrence, leads me to believe that `EPOLLIN` was just noise. Could this be an epoll bug leading to spurious "ghost" events?

**Comment by stefano-garzarella** [comment]
Yes, the `EPOLLIN` event in the middle of the test is strange. To be sure I used `vsockmon` to dump all the packets exchanged between `iperf-vsock` client and server: the server send few bytes at the beginning and at the end of the test. In the middle, it doesn't send any data.

> Could this be an epoll bug leading to spurious "ghost" events?

Could be, should we provide a reproducer without vsock/firecracker involved?
If there is a bug in epoll, should we handle this case handling the EAGAIN properly?

Anyway, how many threads are involved in the epoll() event handling? I'm worried about if we need EPOLLONESHOT or EPOLLEXCLUSIVE flags, but it's not clear to me if there are multiple threads or just one.

However, for now, I still don't understand why we have two different errors for the two cases that should be similar (the data flow is always from the guest to the host).

**Comment by dhrgit** [comment]
> Anyway, how many threads are involved in the epoll() event handling? I'm worried about if we need EPOLLONESHOT or EPOLLEXCLUSIVE flags, but it's not clear to me if there are multiple threads or just one.

All emulation is done on a single thread, which makes this very curious - epoll should've been really simple.

> Could be, should we provide a reproducer without vsock/firecracker involved?

I haven't tried to reproduce it outside Firecracker yet, but it sounds like a good idea.


> If there is a bug in epoll, should we handle this case handling the EAGAIN properly?

Yes, I'm thinking that's the way to go. It does require another change in the state machine, though, which I suspect was the reason your initial patch didn't work: the state machine expects to always be able to yield a packet on receiving `EPOLLIN` (either that, or error out, dropping the connection), so for that case it optimizes out credit updates. So, if no data is read, some credit updates to the guest may get lost, leading to a stalled connection (i.e. the guest locks up waiting for a credit update from Firecracker).


So far, I've tried another approach that seems to work: when receiving `EPOLLOUT | EPOLLIN`, handle only `EPOLLOUT`, and wait to get notified again for any pending `EPOLLIN`, since we're using level-triggered events. The patch is pretty small it seems to work:

```diff
diff --git a/src/devices/src/virtio/vsock/csm/connection.rs b/src/devices/src/virtio/vsock/csm/connection.rs
index 01855924..a49caead 100644
--- a/src/devices/src/virtio/vsock/csm/connection.rs
+++ b/src/devices/src/virtio/vsock/csm/connection.rs
@@ -418,11 +418,6 @@ where

     /// Notify the connection about an event (or set of events) that it was interested in.
     fn notify(&mut self, evset: EventSet) {
-        if evset.contains(EventSet::IN) {
-            // Data can be read from the host stream. Setting a Rw pending indication, so that
-            // the muxer will know to call `recv_pkt()` later.
-            self.pending_rx.insert(PendingRx::Rw);
-        }

         if evset.contains(EventSet::OUT) {
             // Data can be written to the host stream. Time to flush out the TX buffer.
@@ -454,6 +449,11 @@ where
                 self.pending_rx.insert(PendingRx::CreditUpdate);
             }
         }
+        else if evset.contains(EventSet::IN) {
+            // Data can be read from the host stream. Setting a Rw pending indication, so that
+            // the muxer will know to call `recv_pkt()` later.
+            self.pending_rx.insert(PendingRx::Rw);
+        }
     }
 }
```

**Comment by stefano-garzarella** [comment]
@dhrgit the patch seems to work also on cloud-hypervisor and QEMU + vhost_user_vsock application (based on CH crates).
I run the`iperf3` for a couple of minutes, maybe I need to run it more to be sure.

The strange thing is that now also the `TxBufFull` error disappears, why?

Anyway, the patch seems too hacky for me, are we sure we're not missing any events?

I'll try to write a reproducer to stress the epoll() on UDS. to understand better the epoll() behavior.

**Comment by dhrgit** [comment]
I do agree it's a hacky fix. I'll post a PR with a more permanent solution (i.e. handling `EAGAIN`). No events should get lost, though. Since we're level-triggered, we'll keep getting notified, until the generating condition goes away. The hacky part, IMO, is that the order in which we handle independent events becomes important (i.e. always handle `EPOLLOUT` before `EPOLLIN`).

Re TX buffer, while I can consistently reproduce the issue with a 4.14 guest (with the above patch applied), that isn't the case for 5.6. I'll need to look deeper into it.

**Comment by stefano-garzarella** [comment]
> The hacky part, IMO, is that the order in which we handle independent events becomes important (i.e. always handle `EPOLLOUT` before `EPOLLIN`).

I agree, and it may be related to some problems in the state machine, but I'm not sure.

I tried to write a simple reproducer to stress epoll() without success, but I'm really curious why we are receiving the EPOLLIN event.

I'll try to find a way to trace code and events, I'm pretty new on Rust, so if you have some pointers, let me know.

> Re TX buffer, while I can consistently reproduce the issue with a 4.14 guest (with the above patch applied), that isn't the case for 5.6. I'll need to look deeper into it.

Thanks!

**Comment by sboeuf** [comment]
@dhrgit @stefano-garzarella just curious, did you make any progress on this issue?

**Comment by gc-plp** [comment]
@sboeuf, I've also started looking into the issue.

I managed to reproduce the broken pipe issue after a few tries, but so far it only happened after about 12 minutes when using the 5.6 kernel.

**Comment by stefano-garzarella** [comment]
@sboeuf I wrote a [test](https://github.com/stefano-garzarella/rust-epoll-test) to stress the epoll(), but I didn't find anything wrong on `Linux 5.6.8-200.fc31.x86_64`.

I used the `epoll` crate that should be the same of CloudHypervisor.

@gc-plp let me know if i can help in some way.

**Comment by gc-plp** [comment]
I tested the issue with a few setups and changed following error handler in `connection.rs` to:
```
diff --git a/src/devices/src/virtio/vsock/csm/connection.rs b/src/devices/src/virtio/vsock/csm/connection.rs
index 01855924..ccc7308d 100644
--- a/src/devices/src/virtio/vsock/csm/connection.rs
+++ b/src/devices/src/virtio/vsock/csm/connection.rs
@@ -241,10 +241,12 @@ where
             Err(err) => {
                 // We are not expecting any errors when reading from the underlying stream. If
                 // any show up, we'll immediately kill this connection.
-                error!(
-                    "vsock: error reading from backing stream: lp={}, pp={}, err={:?}",
-                    self.local_port, self.peer_port, err
-                );
+                if err.kind() == std::io::ErrorKind::WouldBlock {
+                    // If reading from the host socket would've blocked now, it might succeed later.
+                    return Err(VsockError::NoData);
+                }
+
                 pkt.set_op(uapi::VSOCK_OP_RST);
             }
         };
```

When using the patch above, even though `NoData` is occasionally returned from the `recv` call, the throughput doesn't always decrease to `0 bits/sec`, which could indicate that there is a race condition somewhere.

**Comment by stefano-garzarella** [comment]
@gc-plp when I found the [issue in CloudHypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1001), I tried exactly the same patch as I wrote in the CH issue, but I had the stall every time `NoData` is returned.

**Comment by gc-plp** [comment]
@stefano-garzarella, I also noticed slight differences when linking the Firecracker binary against musl/ glibc, with the musl build showing more consistent failures.

**Comment by sboeuf** [comment]
@stefano-garzarella we recently added musl support to Cloud-Hypervisor, so this might help reproducing the issue as well.

**Comment by stefano-garzarella** [comment]
@gc-plp do you mean more frequently? Do you have an idea why? I'm not sure if it is related to virtio-vsock credit mechanism, I'll check packet exchanged in the guest using `tcpdump`.

@sboeuf thanks for the info! I'll try, but also with glibc I had stall very frequently.

**Comment by gc-plp** [comment]
@stefano-garzarella yes, it seems that for some reason when using musl it triggers more frequently.

I was basically trying to figure out if using different setups has a higher incidence of triggering the issue, so I ran about 10 iterations of starting two Firecracker 0.21.1 VMs in parallel and started `iperf-vsock` in the host as a server. In 8 out of the 10 runs, the iperf instance in the musl build stopped generating traffic long before the gnu build - less than 30s vs. about 5 minutes.

Most probably both libc libraries work fine, but for some reason the musl build makes it easier to trigger it.

**Comment by gc-plp** [comment]
Hi @stefano-garzarella. We've addressed some of the issues in [PR #1910](https://github.com/firecracker-microvm/firecracker/pull/1910).

I've been stress testing this version over the weekend and it seems stable now. 
Could you pull the updates on your side and check if the original problem still reproduces?

Wwe still found corner cases where the device seems to hang, such as when using larger TX buffers, combined with a credit update threshold set to 50% of the buffer:
```diff
diff --git a/src/devices/src/virtio/vsock/csm/mod.rs b/src/devices/src/virtio/vsock/csm/mod.rs
index 2a22c3e2..ef511b1b 100644
--- a/src/devices/src/virtio/vsock/csm/mod.rs
+++ b/src/devices/src/virtio/vsock/csm/mod.rs
@@ -10,11 +10,11 @@ pub use connection::VsockConnection;
·
 pub mod defs {
     /// Vsock connection TX buffer capacity.
-    pub const CONN_TX_BUF_SIZE: usize = 64 * 1024;
+    pub const CONN_TX_BUF_SIZE: usize = 4 * 64 * 1024;
·
     /// After the guest thinks it has filled our TX buffer up to this limit (in bytes), we'll send
     /// them a credit update packet, to let them know we can handle more.
-    pub const CONN_CREDIT_UPDATE_THRESHOLD: usize = CONN_TX_BUF_SIZE - 4 * 4 * 1024;
+    pub const CONN_CREDIT_UPDATE_THRESHOLD: usize = CONN_TX_BUF_SIZE - 2 * 64 * 1024;
·
     /// Connection request timeout, in millis.
     pub const CONN_REQUEST_TIMEOUT_MS: u64 = 2000;
```

While looking through the kernel vsock code, we found [this](https://github.com/torvalds/linux/blob/master/net/vmw_vsock/virtio_transport_common.c#L378) comment, stating that occasional transmission stalls can be seen.

Do you have any insight on why this happens? Should the buffers that Firecracker creates be set to specific values to avoid the issue described above?

I managed to make the connection hang with a 5.6 kernel.

**Comment by stefano-garzarella** [comment]
> Hi @stefano-garzarella. We've addressed some of the issues in [PR #1910](https://github.com/firecracker-microvm/firecracker/pull/1910).
> 
> I've been stress testing this version over the weekend and it seems stable now.
> Could you pull the updates on your side and check if the original problem still reproduces?

Hi @gc-plp,
Thanks! I'll try next days!

> 
> While the PR seems to work, we still found corner cases where the device hangs, such as when using larger TX buffers, combined with a credit update threshold set to 50% of the buffer:
> 
> ```diff
> diff --git a/src/devices/src/virtio/vsock/csm/mod.rs b/src/devices/src/virtio/vsock/csm/mod.rs
> index 2a22c3e2..ef511b1b 100644
> --- a/src/devices/src/virtio/vsock/csm/mod.rs
> +++ b/src/devices/src/virtio/vsock/csm/mod.rs
> @@ -10,11 +10,11 @@ pub use connection::VsockConnection;
> ·
>  pub mod defs {
>      /// Vsock connection TX buffer capacity.
> -    pub const CONN_TX_BUF_SIZE: usize = 64 * 1024;
> +    pub const CONN_TX_BUF_SIZE: usize = 4 * 64 * 1024;
> ·
>      /// After the guest thinks it has filled our TX buffer up to this limit (in bytes), we'll send
>      /// them a credit update packet, to let them know we can handle more.
> -    pub const CONN_CREDIT_UPDATE_THRESHOLD: usize = CONN_TX_BUF_SIZE - 4 * 4 * 1024;
> +    pub const CONN_CREDIT_UPDATE_THRESHOLD: usize = CONN_TX_BUF_SIZE - 2 * 64 * 1024;
> ·
>      /// Connection request timeout, in millis.
>      pub const CONN_REQUEST_TIMEOUT_MS: u64 = 2000;
> ```
> 
> While looking through the kernel vsock code, we found [this](https://github.com/torvalds/linux/blob/master/net/vmw_vsock/virtio_transport_common.c#L378) comment, stating that occasional transmission stalls can be seen.

The comment says that a stall is expensive, but it is time limited. This means that if we set a low threshold, the transmitter can stall waiting the credit update message, also if there is a space in the remote buffer, but it shouldn't happen that it stalls forever.

> 
> Do you have any insight on why this happens? Should the buffers that Firecracker creates be set to specific values to avoid the issue described above?

I think some problems could be in the `peer_needs_credit_update()`.

1. You are not taking care of the remote buffer size (`peer_buf_alloc`), but you're using a local constant, and it should be wrong.
   Please take a look at https://github.com/torvalds/linux/blob/master/net/vmw_vsock/virtio_transport_common.c#L374

2. u32 wrapping (I'm not a Rust expert, sorry if I'm saying something wrong) of `fwd_cnt` and `last_fwd_cnt`.
   In the kernel we have this formula to calculate the free space:
   ```c
   free_space = vvs->buf_alloc - (vvs->fwd_cnt - vvs->last_fwd_cnt);
   ...
   if (free_space < VIRTIO_VSOCK_MAX_PKT_BUF_SIZE) {
      virtio_transport_send_credit_update(...)
   }
   ```
   `fwd_cnt` is always grater than `last_fwd_cnt`, but a wrap around will happen and in C it should not be a problem, it continues to work. For example:
   ```c
   fwd_cnt = 4073;
   last_fwd_cnt = 4294967290;
   /* fwd_cnt - last_fwd_cnt = 4079 */
   ```
   Are we sure that is working also in Rust?

**Comment by dhrgit** [comment]
> Are we sure that is working also in Rust?

We use `std::num::Wrapping` in Rust for this purpose, which disables overflow checks, effectively acting like C.

I don't think the problem lies in how we handle flow control, though. Everything appears to work fine flow-control-wise, up until we get the first unexpected write error (that unexplained `EAGAIN` after `EPOLLOUT`). At the same time, the guest also seems to stop sending data. Apparently this is not due to it waiting on a credit update, since forcing a credit update right after getting the write error still doesn't unblock the guest.

This leads me to suspect that these two events (the unexpected write error and the guest sopping TX) are both caused by some other, yet unknown, exceptional condition.

**Comment by stefano-garzarella** [comment]
> > Are we sure that is working also in Rust?
> 
> We use `std::num::Wrapping` in Rust for this purpose, which disables overflow checks, effectively acting like C.

@dhrgit thanks for the explanation!

> I don't think the problem lies in how we handle flow control, though. Everything appears to work fine flow-control-wise, up until we get the first unexpected write error (that unexplained `EAGAIN` after `EPOLLOUT`). At the same time, the guest also seems to stop sending data. Apparently this is not due to it waiting on a credit update, since forcing a credit update right after getting the write error still doesn't unblock the guest.
> 
> This leads me to suspect that these two events (the unexpected write error and the guest sopping TX) are both caused by some other, yet unknown, exceptional condition.

If even forcing a credit update does not unblock the guest, I agree that could be a side effect. I'll try putting probes (or using tcpdump) in the guest kernel to see why it stops sending packets.

**Comment by stefano-garzarella** [comment]
@gc-plp @dhrgit finally I backported the patches on top of my Cloud Hypervisor [branch](https://github.com/stefano-garzarella/cloud-hypervisor/tree/vhost-user-vsock) with the vhost-user-vsock application that I'm using with QEMU .

I discovered this issue running iperf inside a VM handled by QEMU with vhost-user-vsock-pci device and vhost-user-vsock application made with Cloud Hypervisor/Firecracker vsock crates.

So I tried again this test and after about ~140 seconds the guest stops to send bytes.
I restarted iperf in the guest (without reboot the VM) and it sends again bytes but it stops after ~15 seconds. I restarted again and it lasted for ~70 seconds.
My host is Fedora 31 (`5.6.13-200.fc31.x86_64`).

At this point the virtqueues seem handled well and the issue should be related to the single connection, so tomorrow I'll try to investigate something in the guest to figure out why it stops (if it stops).

**Comment by gc-plp** [comment]
@stefano-garzarella I noticed similar issues on my end when testing with the increased buffer sizes - the moment when the data stops being sent is not predictable.
Are you using the 64k buffer size?

Also, are you using the latest patch from PR #1910?

**Comment by stefano-garzarella** [comment]
> @stefano-garzarella I noticed similar issues on my end when testing with the increased buffer sizes - the moment when the data stops being sent is not predictable.
> Are you using the 64k buffer size?

@gc-plp yes, I used the default `pub const CONN_TX_BUF_SIZE: u32 = 64 * 1024;`

> 
> Also, are you using the latest patch from PR #1910?

I backported the following patches on my [CH branch](https://github.com/stefano-garzarella/cloud-hypervisor/tree/vhost-user-vsock):

```
665ac10c vsock: flow control fix
c535624a vsock: fixed rxq logic
```

**Comment by gc-plp** [comment]
Does the issue still reproduce when running in reverse mode? In the comment [here](https://github.com/firecracker-microvm/firecracker/issues/1751#issuecomment-609705772), you mentioned that running iperf in reverse mode, with the server being hosted on the host is hanging. 
I'm just curious on what usecases are still causing problems :)

Also, is it possible to share some data on your hardware and what the average throughput before the connection stalls is?
I noticed that there are significant differences when running on different CPUs. For example an Intel Gen 7 CPU seems to have almost twice the throughput of an Intel Gen 6 CPU.

**Comment by stefano-garzarella** [comment]
> Does the issue still reproduce when running in reverse mode? In the comment [here](https://github.com/firecracker-microvm/firecracker/issues/1751#issuecomment-609705772), you mentioned that running iperf in reverse mode, with the server being hosted on the host is hanging.
> I'm just curious on what usecases are still causing problems :)

The test that I did in the previous comment was the following:

```
guest$ ./iperf3 --vsock -c 2 -t 1000
Connecting to host 2, port 5201
[  5] local 4 port 378235923 connected to 2 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   106 MBytes   885 Mbits/sec                  
[  5]   1.00-2.00   sec   655 MBytes  5.50 Gbits/sec                  
[  5]   2.00-3.00   sec   899 MBytes  7.54 Gbits/sec
...
[  5] 163.00-164.00 sec   843 MBytes  7.07 Gbits/sec                  
[  5] 164.00-165.00 sec   236 MBytes  1.98 Gbits/sec                  
[  5] 165.00-166.00 sec  0.00 Bytes  0.00 bits/sec                  
[  5] 166.00-167.00 sec  0.00 Bytes  0.00 bits/sec
```

```
host$ ./iperf3 --vsock -s -B /tmp/vm4.vsock
```

I just tried the same test of the [comment](https://github.com/firecracker-microvm/firecracker/issues/1751#issuecomment-609705772) which you referred (server in the guest and client in the host running in reverse mode) and surprisingly it always hangs in the first second of execution in all test that I did:

```
guest$ ./iperf3 --vsock -s
```

```
host$ ./iperf3 --vsock -c /tmp/vm4.vsock -R -t 10000
Connecting to host /tmp/vm4.vsock, port 5201
Reverse mode, remote host /tmp/vm4.vsock is sending
[  5] local :: port 0 connected to 6d34:2e76:736f:636b:: port 12148
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  1.31 MBytes  11.0 Mbits/sec                  
[  5]   1.00-2.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec                  
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec 
```

In both cases the guest is sending and the host is receiving. I also tried the opposite (host sends, guest receives) running the client in the host, or in the guest in reverse mode, and it works well, but the throughput is under 0.9 Gbps.

> Also, is it possible to share some data on your hardware 

```
CPU: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
RAM: DDR4 (2133 MT/s)
```

> and what the average throughput before the connection stalls is?

Around 7 Gbps, but in `reverse mode` it hangs immediately.

**Comment by stefano-garzarella** [comment]
@gc-plp @dhrgit  in `pub fn flush_to<W>(&mut self, sink: &mut W) -> Result<usize>`, at the end of the function, we recursively call `flush_to`.

What happen if the second write fails?
IIUC the `tail` is increased after the first `write` but we are not increasing `fwd_cnt`, because it will generate an error.

**Comment by gc-plp** [comment]
@stefano-garzarella at one point during the investigation, I rewrote that function to only be called once, but I still hit the stalling issue. But perhaps at that moment, I was hitting one of the previously fixed problems and eventually reverted it.

It's a good catch to check its logic. I'll look into it.

**Comment by gc-plp** [comment]
@stefano-garzarella, I pushed a new PR version which seems to work so far.

I started a test with all the combinations that you also ran - `host to guest` and `guest to host` in both reverse and normal mode:
```
######### host2guest_normal #########
# guest
iperf3 --vsock -t 86400 -c 2

# host
./iperf3 --vsock -s -B v.sock

######### host2guest_reverse #########
# guest
iperf3 --vsock -t 86400 -c 2 -R

# host
./iperf3 --vsock -s -B v.sock

######### guest2host_normal #########
# guest
iperf3 --vsock -s

# host
./iperf3 --vsock -t 86400 -c v.sock

######### guest2host_reverse #########
# guest
iperf3 --vsock -s

# host
./iperf3 --vsock -t 86400 -c v.sock -R
```

The microVMs have been running for a while on a Xeon 8175M based server without any crashes or stalls, but as I was mentioning before, there seem to be more frequent problems when running on processors that have better per-core performance than this one.

Could you try these patches on your side?

**Comment by stefano-garzarella** [comment]
@gc-plp I saw the flash_to() fix, now it seems correct to me, but unfortunately I still have problem here with QEMU + Cloud Hypervisor's vhost_user_vsock application, maybe it is unrelated and it is another problem on my side :-(

```
guest$ ./iperf3 --vsock -s
```

```
host$ ./iperf3 --vsock -c /tmp/vm4.vsock -R -t 10000
Connecting to host /tmp/vm4.vsock, port 5201
Reverse mode, remote host /tmp/vm4.vsock is sending
[  5] local :: port 0 connected to 6d34:2e76:736f:636b:: port 12148
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   137 MBytes  1.15 Gbits/sec                  
[  5]   1.00-2.00   sec   875 MBytes  7.34 Gbits/sec                  
[  5]   2.00-3.00   sec  1.01 GBytes  8.66 Gbits/sec                  
[  5]   3.00-4.00   sec  1.02 GBytes  8.74 Gbits/sec                  
[  5]   4.00-5.00   sec  1008 MBytes  8.46 Gbits/sec  
...
[  5] 119.00-120.00 sec   446 MBytes  3.74 Gbits/sec                  
[  5] 120.00-121.00 sec  0.00 Bytes  0.00 bits/sec                  
[  5] 121.00-122.00 sec  0.00 Bytes  0.00 bits/sec                  
[  5] 122.00-123.00 sec  0.00 Bytes  0.00 bits/sec                  
[  5] 123.00-124.00 sec  0.00 Bytes  0.00 bits/sec 
```

In the vhost-user-vsock application I had these logs:

```
[2020-06-04T10:38:04Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[2020-06-04T10:38:06Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[2020-06-04T10:38:10Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[2020-06-04T10:38:15Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[2020-06-04T10:38:15Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
[2020-06-04T10:38:17Z WARN  vm_virtio::vsock::csm::connection] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=1073741827, pp=5201, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
```

I'm trying with Firecracker but for now it isn't hanging (after about 1700 seconds), also if I'm seeing "vsock: unexpected EWOULDBLOCK while reading from backing stream" warnings.
The main difference is the throughput, with QEMU + vhost-user-vsock I'm reaching 8.5 Gbps, with Firecracker I'm always under 5 Gbps (Guest -> Host)

**Comment by gc-plp** [comment]
In the test bench that I started, Firecracker uses an increased buffer size which reaches about 40 Gbits/sec:

```
diff --git a/src/devices/src/virtio/vsock/csm/mod.rs b/src/devices/src/virtio/vsock/csm/mod.rs
index 57d9440a..6f73feec 100644
--- a/src/devices/src/virtio/vsock/csm/mod.rs
+++ b/src/devices/src/virtio/vsock/csm/mod.rs
@@ -10,7 +10,7 @@ pub use connection::VsockConnection;
 
 pub mod defs {
     /// Vsock connection TX buffer capacity.
-    pub const CONN_TX_BUF_SIZE: u32 = 64 * 1024;
+    pub const CONN_TX_BUF_SIZE: u32 = 64 * 1024 * 1024;
 
     /// When the guest thinks we have less than this amount of free buffer space,
     /// we will send them a credit update packet.
```

Related to the warnings, I'm also seeing them, but not too often. 

Does increasing the buffer size trigger the bug quicker?

**Comment by stefano-garzarella** [comment]
> Does increasing the buffer size trigger the bug quicker?

@gc-plp yes, with this chage the `vhost_user_vsock` application prints a lot of warnings and after ~10 sec the guest hangs.

**Comment by sboeuf** [comment]
@stefano-garzarella have you ported the fix from @gc-plp to your Cloud Hypervisor's vhost_user_vsock application?
I think we also need to port the fix to Cloud-Hypervisor virtio-vsock device.

**Comment by stefano-garzarella** [comment]
@sboeuf yes, for testing the vhost-user-vsock application.
When the patches will be merged, I'll send a PR to Cloud-Hypervisor.

**Comment by gc-plp** [comment]
Keeping it open until the Cloud Hypervisor issue is fixed.

**Comment by stefano-garzarella** [comment]
I just backported the patches in Cloud Hypervisor and the issue seems fixed: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1343

**Comment by stefano-garzarella** [comment]
Just as a note, the PR fixes the issue with Cloud Hypervisor, but I still have some stalls with QEMU + vhost-user-vsock, maybe it could be related to some mistakes that I did while implementing vhost-user-vsock, I'll investigate better.

**Comment by gc-plp** [comment]
@stefano-garzarella sounds like we can close this issue for now, thanks for all your help :)

**Comment by jackloomen** [comment]
Kernel 5.10.43. I'm getting this problem on firecracker (latest release) when using waypipe (wayland over network).
It does push a lot of bandwidth (uncompressed visuals).

Results in broken pipes for socat and these:

```
2021-07-15T23:42:32.073125943 [anonymous-instance:main:WARN:src/devices/src/virtio/vsock/csm/connection.rs:240] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=2000, pp=593102777, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
2021-07-15T23:42:36.278132217 [anonymous-instance:main:WARN:src/devices/src/virtio/vsock/csm/connection.rs:240] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=2000, pp=593102777, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
2021-07-15T23:42:40.359824045 [anonymous-instance:main:WARN:src/devices/src/virtio/vsock/csm/connection.rs:240] vsock: unexpected EWOULDBLOCK while reading from backing stream: lp=2000, pp=593102777, err=Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }
```

**Comment by gc-plp** [comment]
@jackloomen, what Firecracker version are you using? If you're using the latest release, could you try building from master?
